sdk/cxx/include/sawtooth/address_mapper.h

```diff
@@ -0,0 +1,88 @@
+/* Copyright 2017 Intel Corporation
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+------------------------------------------------------------------------------*/
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "sawtooth/message_stream.h"
+
+namespace sawtooth {
+
+// Raised if any of the addresses generated violate
+// the merkel trie address constraints.
+class AddressFormatError: public std::runtime_error {
+ public:
+    explicit AddressFormatError(std::string const& error)
+        : std::runtime_error(error)
+    {}
+};
+
+
+
+// Helper class to provide mappings from domain identifiers
+// to merkle trie addresses. This class is designed to be overloaded, such
+// that implementers of TransactionProcessors can define their own mappings
+// from their native object ids to merkle trie addresses.
+//
+// The default implementation uses a SHA512 hash to map both the key and the
+// namespace to the mekle trie address. Collisions in the key mappings should
+// be uncommon but still possible, data storage behind the keys.
+//
+// Each instance of the mapper holds a namespace that has is mapped to a
+// namespace prefix when it is constructed. If your transaction processor
+// operates on multiple namespaces, it is recommened that you use an instance
+// for each namespace.
+class AddressMapper {
+ public:
+    // Constructor for an aaddress mapping object, Takes the
+    // namesspace it will be mapping objects to as a parameter.
+    explicit AddressMapper(const std::string& namespace_);
```

Comment:
Reading the comment I wasn't sure if this took "intkey" or "1cf126". From the source it looks like "intkey".